### ISSUE TITLE
Codegen: sign function in Fortran

### ISSUE THREAD

**Issue by Franjov** [sympy/sympy#12267]
The Fortran code generated by Sympy for the sign function is not a valid Fortran syntax.

With Sympy 1.0 and Python 3.6:
```python
In [1]: import sympy as sp

In [2]: from sympy.abc import x

In [3]: sp.fcode(sp.sign(x))
Out[3]: '      sign(x)'
```
(The same behavior is obtained with `sympy.utilities.codegen`.)

The syntax of the sign function in Fortran is given [here](https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html).

I guess it would be preferable to generate something like `sign(1, x)` or `sign(1d0, x)` depending of the context. I'm not sure how to get the same behavior than Sympy for `x=0`, but at least it would be compilable.

**Comment by asmeurer** [comment]
The x=0 case probably matters, because SymPy could return an expression that uses `sign` assuming that meaning. Does Fortran have ternary expressions?

**Comment by bjodah** [comment]
Looks like fortran does support ternary expressions though `merge`:
https://en.wikipedia.org/wiki/%3F:#Fortran

**Comment by mamakancha** [comment]
@bjodah 
I would like to work in this issue. Could you please help me out?
Is the following code correct?
`sp.sign(x) = merge(0, sp.sign(x), x is 0)`

**Comment by bjodah** [comment]
@mamakancha no not quite. Fortran knows nothing about `sp.`. You can look at the C version:
```
$ grep -A1 --include "*.py" _print_sign -R sympy/
```

**Comment by SatyaPrakashDwibedi** [comment]
@**bjodah** I think the changes should be done in `doprint()` of codeprinter.py` but not quite sure where and what the code should be .

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah I think the changes should be done in doprint() of codeprinter.py` but not quite sure where and what the code should be ; some help is required.

**Comment by SatyaPrakashDwibedi** [comment]
@**bjodah** I think the changes should be done in `doprint()` of codeprinter.py but not quite sure where and what the code should be

**Comment by bjodah** [comment]
@SatyaPrakashDwibedi I don't think that's needed. Simply adding a method to the printer class for fortran code should be enough (in analogy with the `_print_sign` method in the `CCodePriner` class).

**Comment by Franjov** [comment]
With a recent version a `gfortran`, I can confirm that
```fortran
merge(0, sign(1, x), x == 0)
```
```fortran
merge(0e0, sign(1e0, x), x == 0e0)
```
and
```fortran
merge(0d0, sign(1d0, x), x == 0d0)
```
work as expected, for respectively integer, real and double precision `x`.

Note that [`merge`](https://gcc.gnu.org/onlinedocs/gfortran/MERGE.html) is only defined for Fortran 95 standard (or later). Another solution might be needed for the legacy Fortran 77 standard.

**Comment by bjodah** [comment]
Yes, I noticed that `merge` is a rather "new" addition. I don't know of an easy fix for fortran 77, there I think one would have to implement it using if statements. But supporting "only" F95+ is still a step in the right direction.

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah will this work?
`>>> def p(func):
...   if func.args[0].is_positive:
...     return '(SIGN(1,{0})'.format((func.args[0]))
...   elif func.args[0].is_negative:
...     return '(SIGN(-1,{0})'.format((func.args[0]))
...   else:
...     return '(SIGN(0,{0})'.format((func.args[0]))
...`

**Comment by bjodah** [comment]
@SatyaPrakashDwibedi no that will not work. The argument might be symbolic and its sign is not known prior to runtime (or we would obviously not need to generate any call at all). You should look at `test_fcode.py` and write a test for sign checking that generated code looks like @Franjov suggested, then run ./bin/test sympy/printing/tests/test_fcode.py` and watch it fail. Then edit `sympy/printing/fcode.py` to give the code printer a printing method analogous to the one in `ccode.py`. If done right the test should now pass. You should then verify that generated code actually compiles using e.g. `gfortran`. Then just commit and push and open a PR.

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah Will this work for test_fcode.py
`def test_sign():`
    ` x=symbols('x',Integer=True)`
 `    y=symbols('y',real=True)`
 `    z=symbols('z',Double=True)`
 `    assert fcode(sign(x))=="      merge(0, sign(1, x), x == 0)"`
 `    assert fcode(sign(y))=="      merge(0e0, sign(1e0, x), x == 0e0)"`
 `    assert fcode(sign(z))=="      merge(0d0, sign(1d0, x), x == 0d0)" `

**Comment by bjodah** [comment]
@SatyaPrakashDwibedi, that looks quite good, I'd suggest some minor changes:
```
def test_sign():
    x=symbols('x')
    y=symbols('y', integer=True)
    assert fcode(sign(x), source_format='free') == "merge(0d0, dsign(1d0, x), x == 0d0)"
    assert fcode(sign(y), source_format='free') == "merge(0, isign(1, x), x == 0)"
```
maybe we should handle more cases, but I don't write that much fortran so it's hard for me to be sure, @Franjov what do you think?

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah Should I move to `fcode.py`

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah I think there should be a check in `fcode.py` for type of variables to return right type of string.

**Comment by Franjov** [comment]
You could also deal with the complex case. Something like
```fortran
merge( cmplx(0d0, 0d0), z/abs(z), z == cmplx(0d0, 0d0))
```
(not tested in Fortran yet).

Actually, it is probably safer to code `x == 0d0` as something like `abs(x) < epsilon`.
But `epsilon` might be too dependent of the problem for a code generator without more context.

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah as stated by @Franjov I edited t
test_fcode as:
```
def test_sign():
    x=symbols('x')
    y=symbols('y', integer=True)
    z=symbols('z', complex=True)
    assert fcode(sign(x), source_format='free') == "merge(0d0, dsign(1d0, x), x == 0d0)"
    assert fcode(sign(y), source_format='free') == "merge(0, isign(1, y), y == 0)"
    assert fcode(sign(z), source_format='free') == "merge( cmplx(0d0, 0d0), z/abs(z), abs(z) == 0d0)"
```
and in fcode
```
def _print_sign(self,func):
        if func.args[0].is_integer:
            return "merge(0, isign(1,{0}), {0}==0)".format(self._print(func.args[0]))
        elif func.args[0].is_complex:
            return "merge(cmplx(0d0, 0d0), {0}/abs({0}), abs({0}) == 0d0)".format(self._print(func.args[0]))
        else:
            return "merge(merge(0d0, dsign(1d0, {0}), {0} == 0d0)".format(self._print(func.args[0]))

```
which gives output as:
```
>>> x=symbols('x',integer=True)
>>> y=symbols('y',complex=True)
>>> z=symbols('z')
>>> fcode(sign(x))
'      merge(0, isign(1,x), x==0)'
>>> fcode(sign(y))
'      merge(cmplx(0d0, 0d0), y/abs(y), abs(y) == 0d0)'
>>> fcode(sign(z))
'      merge(merge(0d0, dsign(1d0, z), z == 0d0)'
```

**Comment by bjodah** [comment]
@SatyaPrakashDwibedi that looks great! Commit your changes in a new branch, push that branch to your sympy repo on github and create a pull request from there, remember to reference this issue in the PR. Feel free to ping me in the PR for review after tests are passing or for guidance.

Regarding the complex case, @Franjov that would give us a complex unit vector unless z=0 + 0j
I could also imagine people expecting sign on a complex operate on the real part and the imaginary part separately. What do you think? Regarding using a small value "epsilon" instead of checking for equality to 0: the beauty with the code printers is that a user can quite eaisly override the print method by subclassing in order to get that behavior.

**Comment by Franjov** [comment]
@bjodah 
About the complex sign function: I agree with you, but that's actually the behavior of [Sympy's sign function](http://docs.sympy.org/dev/modules/functions/elementary.html#sign).

@SatyaPrakashDwibedi 
Look out for the superfluous `merge` in the last case.

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah  I made the pull request

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah why it is giving assertion error?

**Comment by Franjov** [comment]
Maybe the difference of spacing in the generated code between `print_sign` and `test_sign`?

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov It seems you are right; when I deleted all the spacings there was no error .
Are the spacings required?

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov seems you are right there was no error after removing the spaces

**Comment by Franjov** [comment]
Sorry to dig up the thread, I have a small suggestion:

While looking at the doc, I stumbled upon the `_print_Piecewise` method.
It uses `merge` for modern Fortran like us, but can also use an `if`construct for older code.
```python
>>> print(fcode(Piecewise((x,x<1),(x**2,True)), assign_to="var"))
      if (x < 1) then
        var = x
      else
        var = x**2
      end if
```
Wouldn't it be interesting to do the same for the `sign` function? Maybe by relying on `_print_Piecewise`?

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov What should be the code for sign in older version of Fortran

**Comment by Franjov** [comment]
It could be something like
```fortran
      if (x == 0) then
        var = 0
      else
        var = isign(1, x)
      end if
```
for integers,
```fortran
      if (x == 0d0) then
        var = 0d0
      else
        var = dsign(1d0, x)
      end if
```
for real numbers and
```fortran
      if (abs(x) == 0d0) then
        var = cmplx(0d0, 0d0)
      else
        var = x/abs(x)
      end if
```
for complex numbers.
Of course, as for `Piecewise`, it can only work with an `assign_to` argument.

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov I will fix it

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah how should I proceed with this?
while using `assign_to` it returns`'var=None'`
because assignment assigns var to mathematical expression but here `sign` takes only one argument.
So what is your suggestion?

**Comment by Franjov** [comment]
I'm not sure to understand what you mean.
If there is no `assign_to` argument, you can't generate a Fortran 77 code for `sp.sign`. If there is, it should return the code as in my previous message where `var` is the string passed to the `assign_to` argument.
Try for instance: 
```python
sp.fcode(sp.Piecewise((0.0, x < 0.0), (1.0, True)), assign_to='var')
```
Looking at `_print_Piecewise` might be useful to code the same behavior for the sign function.

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah @Franjov 
In `_print_piecewise()` https://github.com/SatyaPrakashDwibedi/sympy/blob/master/sympy/printing/fcode.py#L149
this line is used for assigning expression to var(passed argument)
but if I do in `_print_sign` (for demo)
```
x=S('x')
if func.has(Assignment):
            if func.args[0].is_integer:
                lines=[]
                lines.append(self._print(x))
            return "\n".join(lines)
```
it is giving `'      var = None'` as output.
I am not getting where the mistake is.

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov are `isign(1,x)` `cmplx(0d0,0d0)` are FORTRAN inbuilt functions?

**Comment by Franjov** [comment]
@SatyaPrakashDwibedi 
Yes. See : https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html#SIGN and https://gcc.gnu.org/onlinedocs/gfortran/CMPLX.html#CMPLX

**Comment by SatyaPrakashDwibedi** [comment]
@Franjov check out :
```
>>> x=symbols('x',integer=True)
>>> y=symbols('y',complex=True)
>>> z=symbols('z')
>>> print fcode(sign(x),assign_to="var")
      if (x == 0d0) then
         var = 0(0d0,0d0)
      else
         var = isign(1.0,x)
      end if
>>> print fcode(sign(y),assign_to="var")
      if (y == 0d0) then
         var = cmplx(0d0,0d0)
      else
         var = y/abs(y)
      end if
>>> print fcode(sign(z),assign_to="var")
      if (z == 0d0) then
         var =0d0
      else
         var = dsign(1d0,z)
      end if
>>> print fcode(sign(x),standard=95)
      merge(0, isign(1, x), x == 0)
>>> print fcode(sign(y),standard=95)
      merge(cmplx(0d0, 0d0), y/abs(y), abs(y) == 0d0
>>> print fcode(sign(z),standard=95)
      merge(0d0, dsign(1d0, z), z == 0d0)
```

**Comment by Franjov** [comment]
The output looks good!
You can also check the case with neither `assign_to` nor `standard=95`, and the case with both.

**Comment by bemichel** [comment]
Hi,
I wonder if there is not a problem between `real` and `complex` definition and the code for printing `sign` in fortran.

Indeed, next exemple works fine :
 
```python
>>> x=symbols('x')
>>> y=symbols('y',integer=True)
>>> z=symbols('z',complex=True)
>>> print fcode(sign(x),assign_to="var")
      if (x == 0d0) then
         var =0d0
      else
         var = dsign(1d0,x)
      end if
>>> print fcode(sign(y),assign_to="var")
      if (y == 0) then
         var = 0
      else
         var = isign(1,y)
      end if
>>> print fcode(sign(z),assign_to="var")
      if (abs(z) == 0d0) then
         var = cmplx(0d0,0d0)
      else
         var = z/abs(z)
      end if
>>> print fcode(sign(x),standard=95)
      merge(0d0, dsign(1d0, x), x == 0d0)
>>> print fcode(sign(y),standard=95)
      merge(0, isign(1, y), y == 0)
>>> print fcode(sign(z),standard=95)
      merge(cmplx(0d0, 0d0), z/abs(z), abs(z) == 0d0)`
```   
but it gives some contradiction with the `real` assumption.

As explain in [assumptions ]( https://docs.sympy.org/latest/guides/assumptions.html#predicates), `real` is defined as `complex`.
So if I write this exemple 

```python
>>> import sympy as sp 
>>> x=sp.symbols('x', real=True)
>>> print fcode(sign(x),standard=95)
>>> print(sp.fcode(sp.sign(x),standard=95))                                                                                                                                                                                                                                                                                
      merge(cmplx(0d0, 0d0), x/abs(x), abs(x) == 0d0)
```

The test for `complex` in https://github.com/sympy/sympy/blob/f8bed8c4ee4de92ec2bb9f3076fb45ed407994db/sympy/printing/fortran.py#L206 
does not distinguish `real` from `complex`, but perhaps it is the normal behaviour ?

There is a possibility to distinguish  `real` and `complex` :

```python
>>> a = sp.Symbol("a", real=True)
>>> a.is_real
    True
>>> a.is_complex
    True
>>> b = sp.Symbol("b", complex=True)
>>> b.is_real
>>> bool(b.is_real)
    False
>>> b.is_complex
    True
```
The `_print_sign` could perhaps be written as follow 

```python
    def _print_sign(self, expr):
        from sympy.functions.elementary.complexes import Abs
        arg, = expr.args
        if arg.is_integer:
            new_expr = merge(0, isign(1, arg), Eq(arg, 0))
        elif ((arg.is_complex and not bool(arg.is_real)) or arg.is_infinite):
            new_expr = merge(cmplx(literal_dp(0), literal_dp(0)), arg/Abs(arg), Eq(Abs(arg), literal_dp(0)))
        else:
            new_expr = merge(literal_dp(0), dsign(literal_dp(1), arg), Eq(arg, literal_dp(0)))
        return self._print(new_expr)
```
In this case, we retrieve a `double` statment for `real`

```python
>>> import sympy as sp 
>>> x=sp.symbols('x', real=True)
>>> print fcode(sign(x),standard=95)
>>> print(sp.fcode(sp.sign(x),standard=95))                                                                                                                                                                                                                                                                                
       merge(0d0, dsign(1d0, x), x == 0d0)
```
